pkg/operation/botanist/component/logging/fluentoperator/fluent_operator.go

```diff
@@ -0,0 +1,301 @@
+	client    client.Client
+	namespace string
+	values    Values
+}
+
+// NewFluentOperator creates a new instance of Fluent Operator.
+func NewFluentOperator(
+	client client.Client,
+	namespace string,
+	values Values,
+) component.DeployWaiter {
+	return &fluentOperator{
+		client:    client,
+		namespace: namespace,
+		values:    values,
+	}
+}
+
+func (f *fluentOperator) Deploy(ctx context.Context) error {
+	var (
+		registry = managedresources.NewRegistry(kubernetes.SeedScheme, kubernetes.SeedCodec, kubernetes.SeedSerializer)
+
+		serviceAccount = &corev1.ServiceAccount{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: f.namespace,
+				Labels:    getLabels(),
+			},
+			AutomountServiceAccountToken: pointer.Bool(false),
+		}
+		clusterRole = &rbacv1.ClusterRole{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   roleName,
+				Labels: getLabels(),
+			},
+			Rules: []rbacv1.PolicyRule{
+				{
+					APIGroups: []string{"apps"},
+					Resources: []string{"daemonsets", "statefulsets"},
+					Verbs:     []string{"create", "delete", "get", "list", "patch", "update", "watch"},
+				},
+				{
+					APIGroups: []string{""},
+					Resources: []string{"pods"},
+					Verbs:     []string{"get"},
+				},
+				{
+					APIGroups: []string{""},
+					Resources: []string{"secrets", "configmaps", "serviceaccounts", "services", "namespaces"},
+					Verbs:     []string{"create", "delete", "get", "list", "patch", "update", "watch"},
+				},
+				{
+					APIGroups: []string{"fluentbit.fluent.io"},
+					Resources: []string{"fluentbits", "clusterfluentbitconfigs", "clusterfilters", "clusterinputs", "clusteroutputs", "clusterparsers", "collectors", "fluentbitconfigs", "filters", "outputs", "parsers"},
+					Verbs:     []string{"create", "delete", "get", "list", "patch", "update", "watch"},
+				},
+				{
+					APIGroups: []string{"rbac.authorization.k8s.io"},
+					Resources: []string{"clusterrolebindings", "clusterroles"},
+					Verbs:     []string{"get", "list", "watch", "create"},
+				},
+				{
+					APIGroups: []string{"extensions.gardener.cloud"},
+					Resources: []string{"clusters"},
+					Verbs:     []string{"get", "list", "watch"},
+				},
+			},
+		}
+		clusterRoleBinding = &rbacv1.ClusterRoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   name,
+				Labels: getLabels(),
+			},
+			RoleRef: rbacv1.RoleRef{
+				APIGroup: rbacv1.GroupName,
+				Kind:     "ClusterRole",
+				Name:     clusterRole.Name,
+			},
+			Subjects: []rbacv1.Subject{{
+				Kind:      rbacv1.ServiceAccountKind,
+				Name:      serviceAccount.Name,
+				Namespace: serviceAccount.Namespace,
+			}},
+		}
+		deployment = &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      v1beta1constants.DeploymentNameFluentOperator,
+				Namespace: f.namespace,
+				Labels: utils.MergeStringMaps(getLabels(), map[string]string{
+					resourcesv1alpha1.HighAvailabilityConfigType: resourcesv1alpha1.HighAvailabilityConfigTypeController,
```

Comment:
After adding this label I see that fluent-operator Deployment runs with 2 replicas.
But I don't see any leader election config in the deployment. From the logs, it looks like both controllers are doing job and fighting each other. Can you please check this?
If that's the case either enable leader election for fluent-operator - it seems to be supported:

```
% docker run eu.gcr.io/gardener-project/3rd/kubesphere/fluent-operator:v2.2.0 --help

  -leader-elect
    	Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager.
```

Reply:
```diff
diff --git a/pkg/operation/botanist/component/logging/fluentoperator/fluent_operator.go b/pkg/operation/botanist/component/logging/fluentoperator/fluent_operator.go
index b29a9a0b1..4b11d3aa1 100644
--- a/pkg/operation/botanist/component/logging/fluentoperator/fluent_operator.go
+++ b/pkg/operation/botanist/component/logging/fluentoperator/fluent_operator.go
@@ -144,6 +144,47 @@ func (f *fluentOperator) Deploy(ctx context.Context) error {
 				Namespace: serviceAccount.Namespace,
 			}},
 		}
+		role = &rbacv1.Role{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      roleName,
+				Namespace: f.namespace,
+			},
+			Rules: []rbacv1.PolicyRule{
+				{
+					APIGroups: []string{"coordination.k8s.io"},
+					Resources: []string{"leases"},
+					Verbs:     []string{"create"},
+				},
+				{
+					APIGroups: []string{"coordination.k8s.io"},
+					Resources: []string{"leases"},
+					Verbs:     []string{"get", "watch", "update"},
+				},
+				{
+					APIGroups: []string{""},
+					Resources: []string{"events"},
+					Verbs:     []string{"create", "get", "list", "watch", "patch"},
+				},
+			},
+		}
+		roleBinding = &rbacv1.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      role.Name,
+				Namespace: f.namespace,
+			},
+			Subjects: []rbacv1.Subject{
+				{
+					Kind:      rbacv1.ServiceAccountKind,
+					Name:      serviceAccount.Name,
+					Namespace: serviceAccount.Namespace,
+				},
+			},
+			RoleRef: rbacv1.RoleRef{
+				APIGroup: rbacv1.GroupName,
+				Kind:     "Role",
+				Name:     role.Name,
+			},
+		}
 		deployment = &appsv1.Deployment{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      v1beta1constants.DeploymentNameFluentOperator,
@@ -173,8 +214,11 @@ func (f *fluentOperator) Deploy(ctx context.Context) error {
 								Name:            name,
 								Image:           f.values.Image,
 								ImagePullPolicy: corev1.PullIfNotPresent,
-								Args:            []string{"--disable-component-controllers", "fluentd"},
-								Env:             f.computeEnv(),
+								Args: []string{
+									"--leader-elect=true",
+									"--disable-component-controllers=fluentd",
+								},
+								Env: f.computeEnv(),
 								Resources: corev1.ResourceRequirements{
 									Requests: corev1.ResourceList{
 										corev1.ResourceCPU:    resource.MustParse("20m"),
@@ -236,6 +280,8 @@ func (f *fluentOperator) Deploy(ctx context.Context) error {
 		serviceAccount,
 		clusterRole,
 		clusterRoleBinding,
+		role,
+		roleBinding,
 		deployment,
 		vpa,
 	)

```

Reply:
Is everything from https://github.com/gardener/gardener/blob/e3244825da547000d4c89a1f9472348f9dab902a/pkg/operation/botanist/component/logging/fluentoperator/fluent_operator.go#L98-L129 has to be part of the ClusterRole? The ClusterRole would allow the rules in any Namespace, while a Role would allow the rules only in the Role's namespace.
For example I could imagine that the fluent-operator needs to create daemonsets only in the namespace it runs in (the garden namespace).
I would appreciate if you check whether some of these ClusterRole rules can be moved to a Role.
The rule for `statefulsets` seems to be obsolete now after the fluentd controller is disabled?

Reply:
> The rule for statefulsets seems to be obsolete now after the fluentd controller is disabled?

Has to check that, because the fluent-operator v.2.1.0 includes a `collector` resource, which is used to create a Sts from fluent-bit :D